src/main/java/org/elasticsearch/index/engine/EngineConfig.java

```diff
@@ -0,0 +1,305 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.IndexWriterConfig;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.index.analysis.AnalysisService;
+import org.elasticsearch.index.codec.CodecService;
+import org.elasticsearch.index.deletionpolicy.SnapshotDeletionPolicy;
+import org.elasticsearch.index.indexing.ShardIndexingService;
+import org.elasticsearch.index.merge.policy.MergePolicyProvider;
+import org.elasticsearch.index.merge.scheduler.MergeSchedulerProvider;
+import org.elasticsearch.index.settings.IndexSettings;
+import org.elasticsearch.index.settings.IndexSettingsService;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.similarity.SimilarityService;
+import org.elasticsearch.index.store.Store;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.indices.warmer.InternalIndicesWarmer;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class EngineConfig {
+    private final ShardId shardId;
+    private volatile boolean failOnMergeFailure = true;
+    private volatile boolean failEngineOnCorruption = true;
+    private volatile ByteSizeValue indexingBufferSize;
+    private volatile int indexConcurrency = IndexWriterConfig.DEFAULT_MAX_THREAD_STATES;
+    private volatile boolean compoundOnFlush = true;
+
+    private long gcDeletesInMillis = DEFAULT_GC_DELETES.millis();
+
+    private volatile boolean enableGcDeletes = true;
+    private volatile String codecName = DEFAUTL_CODEC;
+    private final boolean optimizeAutoGenerateId;
+
+    private final ThreadPool threadPool;
+
+    private final ShardIndexingService indexingService;
+    private final IndexSettingsService indexSettingsService;
+    @Nullable
+    private final InternalIndicesWarmer warmer;
+    private final Store store;
+    private final SnapshotDeletionPolicy deletionPolicy;
+    private final Translog translog;
+    private final MergePolicyProvider mergePolicyProvider;
+    private final MergeSchedulerProvider mergeScheduler;
+    private final AnalysisService analysisService;
+    private final SimilarityService similarityService;
+    private final CodecService codecService;
+
+    private final Engine.FailedEngineListener failedEngineListener;
+
+
+    public static final String INDEX_INDEX_CONCURRENCY = "index.index_concurrency";
```

Comment:
Can probably just call this `INDEX_CONCURRENCY` instead of `INDEX_INDEX_CONCURRENCY`